src/testing/vortex/faulty_network.zig

```diff
@@ -0,0 +1,583 @@
+        }
+
+        pipe.recv_count = result catch |err| {
+            log.warn("recv error ({d},{d}): {any}", .{
+                pipe.connection.replica_index,
+                pipe.connection.connection_index,
+                err,
+            });
+            return pipe.connection.try_close();
+        };
+        if (pipe.recv_count == 0) {
+            // Zero bytes means EOF.
+            return pipe.connection.try_close();
+        }
+
+        if (pipe.connection.network.faults.lose) |lose| {
+            if (arbitrary.odds(pipe.connection.network.random, lose.percentage, 100)) {
+                log.debug("losing {d} bytes ({d},{d})", .{
+                    pipe.recv_count,
+                    pipe.connection.replica_index,
+                    pipe.connection.connection_index,
+                });
+                pipe.recv();
+                return;
+            }
+        }
+
+        if (pipe.connection.network.faults.corrupt) |corrupt| {
+            if (arbitrary.odds(pipe.connection.network.random, corrupt.percentage, 100)) {
+                switch (pipe.connection.network.random.enumValue(enum { shuffle, zero })) {
+                    .shuffle => {
+                        log.debug("shuffling {d} bytes ({d},{d})", .{
+                            pipe.recv_count,
+                            pipe.connection.replica_index,
+                            pipe.connection.connection_index,
+                        });
+                        pipe.connection.network.random.shuffle(u8, pipe.buffer[0..pipe.recv_count]);
+                    },
+                    .zero => {
+                        log.debug("zeroing {d} bytes ({d},{d})", .{
+                            pipe.recv_count,
+                            pipe.connection.replica_index,
+                            pipe.connection.connection_index,
+                        });
+                        @memset(pipe.buffer[0..pipe.recv_count], 0);
+                    },
+                }
+            }
+        }
+
+        if (pipe.connection.network.faults.delay) |delay| {
+            assert(delay.jitter_ms <= delay.time_ms);
+            const jitter_diff_ms: i32 = pipe.connection.network.random.intRangeAtMost(
+                i32,
+                -@as(i32, @intCast(delay.jitter_ms)),
+                @as(i32, @intCast(delay.jitter_ms)),
+            );
+            const timeout_duration_ns = @as(
+                u63,
+                @intCast(@as(i32, @intCast(delay.time_ms)) + jitter_diff_ms),
+            ) * std.time.ns_per_ms;
+            log.debug("delaying {} ({d},{d})", .{
+                std.fmt.fmtDuration(timeout_duration_ns),
+                pipe.connection.replica_index,
+                pipe.connection.connection_index,
+            });
+            pipe.io.timeout(
+                *Pipe,
+                pipe,
+                on_timeout,
+                &pipe.send_completion,
+                timeout_duration_ns,
+            );
```

Comment:
Does Vortex pull in anything from vsr? Maybe consider making this use timeouts and calling them from `.tick()`, but it's probably fine as-is!